src/selectors/profile.js

```diff
@@ -139,6 +140,103 @@ export const getCounterSelectors = (index: CounterIndex): CounterSelectors => {
   return selectors;
 };
 
+type IPCCorrelationKey = {|
+  tid: number | void,
+  index: number,
+|};
+
+export class IPCCorrelations {
+  _correlations: Map<string, IPCPairData>;
+
+  constructor() {
+    this._correlations = new Map();
+  }
+
+  _makeKey(tid: number | void, index: number) {
+    return [tid, index].toString();
+  }
+
+  set({ tid, index }: IPCCorrelationKey, data: IPCPairData): void {
+    this._correlations.set(this._makeKey(tid, index), data);
+  }
+
+  get({ tid, index }: IPCCorrelationKey): ?IPCPairData {
+    return this._correlations.get(this._makeKey(tid, index));
+  }
+
+  has({ tid, index }: IPCCorrelationKey): boolean {
+    return this._correlations.has(this._makeKey(tid, index));
+  }
+}
+
+export const getIPCCorrelations: Selector<IPCCorrelations> = createSelector(
+  getThreads,
+  threads => {
+    // Create a unique ID constructed from the source PID, destination PID,
+    // message seqno, and message type. FIXME: with a more-unique message
+    // seqno (or other identifier), we could probably simplify this a lot.
+    function makeMarkerID(srcpid, destpid, messageseqno, messagetype) {
+      return [srcpid, destpid, messageseqno, messagetype].toString();
+    }
+
+    // First, construct a mapping of marker IDs to destination markers for
+    // faster lookup.
+    const dstMarkers = new Map();
+    for (const thread of threads) {
+      for (let i = 0; i < thread.markers.length; i++) {
+        const data = thread.markers.data[i];
+        if (!data || data.type !== 'IPC' || data.direction !== 'receiving') {
+          continue;
+        }
+        const key = makeMarkerID(
+          data.otherpid,
+          data.thispid,
+          data.messageseqno,
+          data.messagetype
+        );
+        if (dstMarkers.has(key)) {
+          // FIXME: When we have a more-reliable way of identifying correlated
+          // IPC markers, we won't have to worry about this.
+          console.warn('[squib] duplicate message found', key);
```

Comment:
Oh, right. I forgot about this bit. I'm going to try and remove the need for this entirely if I can get the backend to behave a bit better...

Reply:
I've pretty much convinced myself that this can't ever happen now, unless the profile is corrupted or something changes substantially in Gecko. I think it's worth leaving a warning here just in case, or maybe even upgrading it to an exception...